### PR TITLE
mimir-build-image: Upgrade Go version to 1.23.0-bookworm

### PR DESCRIPTION
**Pull request by alexweav**
#### What this PR does

Upgrades Go version in mimir-build-image.

This required an upgrade to `golangci-lint`. In turn, the upgrade of this introduced a new linter rule for constant format strings (which in rare cases can be a security thing). Fixed all violations of this new linter rule.

#### Which issue(s) this PR fixes or relates to

n/a

#### Checklist

- [ ] Tests updated.
- [ ] Documentation added.
- [ ] `CHANGELOG.md` updated - the order of entries should be `[CHANGE]`, `[FEATURE]`, `[ENHANCEMENT]`, `[BUGFIX]`.
- [ ] [`about-versioning.md`](https://github.com/grafana/mimir/blob/main/docs/sources/mimir/configure/about-versioning.md) updated with experimental features.

### CODE REVIEW

**Comment by alexweav** [comment]
Not ready to merge yet: Awaiting image to be built and bot commit per instructions https://github.com/grafana/mimir/blob/main/docs/internal/how-to-update-the-build-image.md?plain=1

**Review by aknuds1 — COMMENTED** [review]
I think you have to upgrade to the latest version of golangci-lint also. That's probably why the lint CI job failed.

**Comment by alexweav** [comment]
Thanks @aknuds1 :smile:

The remaining lint errors seem to be new/legitimate ones caught by the upgraded linter - i'll post another commit fixing these up tomorrow.

File: pkg/distributor/distributor_ingest_storage_test.go [hunk 1]
```diff
@@ -98,7 +98,7 @@ func TestDistributor_Push_ShouldSupportIngestStorage(t *testing.T) {
 				// Non-retryable error.
 				1: testkafka.CreateProduceResponseError(0, kafkaTopic, 1, kerr.InvalidTopicException),
 			},
-			expectedErr: fmt.Errorf(fmt.Sprintf("%s %d", failedPushingToPartitionMessage, 1)),
+			expectedErr: fmt.Errorf("%s %d", failedPushingToPartitionMessage, 1),
```
Comment (aknuds1) [inline pkg/distributor/distributor_ingest_storage_test.go:101]:
```suggestion
			expectedErr: fmt.Errorf("%s 1", failedPushingToPartitionMessage),
```

File: pkg/distributor/otel.go [hunk 2]
```diff
@@ -90,7 +90,7 @@ func OTLPHandler(
 				protoBodySize, err := util.ParseProtoReader(ctx, reader, int(r.ContentLength), maxRecvMsgSize, buffers, unmarshaler, compression)
 				var tooLargeErr util.MsgSizeTooLargeErr
 				if errors.As(err, &tooLargeErr) {
-					return exportReq, 0, httpgrpc.Errorf(http.StatusRequestEntityTooLarge, distributorMaxOTLPRequestSizeErr{
+					return exportReq, 0, httpgrpc.Errorf(http.StatusRequestEntityTooLarge, "%s", distributorMaxOTLPRequestSizeErr{
```
Comment (aknuds1) [inline pkg/distributor/otel.go:93]:
```suggestion
					return exportReq, 0, httpgrpc.Error(http.StatusRequestEntityTooLarge, distributorMaxOTLPRequestSizeErr{
```

File: pkg/distributor/otel.go [hunk 3]
```diff
@@ -118,7 +118,7 @@ func OTLPHandler(
 				reader = http.MaxBytesReader(nil, reader, int64(maxRecvMsgSize))
 				if _, err := buf.ReadFrom(reader); err != nil {
 					if util.IsRequestBodyTooLarge(err) {
-						return exportReq, 0, httpgrpc.Errorf(http.StatusRequestEntityTooLarge, distributorMaxOTLPRequestSizeErr{
+						return exportReq, 0, httpgrpc.Errorf(http.StatusRequestEntityTooLarge, "%s", distributorMaxOTLPRequestSizeErr{
```
Comment (aknuds1) [inline pkg/distributor/otel.go:121]:
```suggestion
						return exportReq, 0, httpgrpc.Error(http.StatusRequestEntityTooLarge, distributorMaxOTLPRequestSizeErr{
```

File: pkg/distributor/otel.go [hunk 4]
```diff
@@ -137,7 +137,7 @@ func OTLPHandler(
 		// Check the request size against the message size limit, regardless of whether the request is compressed.
 		// If the request is compressed and its compressed length already exceeds the size limit, there's no need to decompress it.
 		if r.ContentLength > int64(maxRecvMsgSize) {
-			return httpgrpc.Errorf(http.StatusRequestEntityTooLarge, distributorMaxOTLPRequestSizeErr{
+			return httpgrpc.Errorf(http.StatusRequestEntityTooLarge, "%s", distributorMaxOTLPRequestSizeErr{
```
Comment (aknuds1) [inline pkg/distributor/otel.go:140]:
```suggestion
			return httpgrpc.Error(http.StatusRequestEntityTooLarge, distributorMaxOTLPRequestSizeErr{
```

File: pkg/distributor/push.go [hunk 5]
```diff
@@ -164,7 +164,7 @@ func handler(
 			if err := parser(ctx, r, maxRecvMsgSize, rb, &req, logger); err != nil {
 				// Check for httpgrpc error, default to client error if parsing failed
 				if _, ok := httpgrpc.HTTPResponseFromError(err); !ok {
-					err = httpgrpc.Errorf(http.StatusBadRequest, err.Error())
+					err = httpgrpc.Errorf(http.StatusBadRequest, "%s", err.Error())
```
Comment (aknuds1) [inline pkg/distributor/push.go:167]:
```suggestion
					err = httpgrpc.Error(http.StatusBadRequest, err.Error())
```

File: pkg/distributor/push_test.go [hunk 6]
```diff
@@ -534,14 +534,14 @@ func TestHandler_ErrorTranslation(t *testing.T) {
 	}{
 		{
 			name:                 "a generic error during request parsing gets an HTTP 400",
-			err:                  fmt.Errorf(errMsg),
+			err:                  fmt.Errorf("%s", errMsg),
```
Comment (aknuds1) [inline pkg/distributor/push_test.go:537]:
```suggestion
			err:                  errors.New(errMsg),
```

File: pkg/distributor/push_test.go [hunk 7]
```diff
@@ -534,14 +534,14 @@ func TestHandler_ErrorTranslation(t *testing.T) {
...
 			expectedHTTPStatus:   http.StatusBadRequest,
 			expectedErrorMessage: errMsg,
 			expectedLogs:         []string{`level=error user=testuser msg="detected an error while ingesting Prometheus remote-write request (the request may have been partially ingested)" httpCode=400 err="rpc error: code = Code(400) desc = this is an error" insight=true`},
 		},
 		{
 			name:                 "a gRPC error with a status during request parsing gets translated into HTTP error without DoNotLogError header",
-			err:                  httpgrpc.Errorf(http.StatusRequestEntityTooLarge, errMsg),
+			err:                  httpgrpc.Errorf(http.StatusRequestEntityTooLarge, "%s", errMsg),
```
Comment (aknuds1) [inline pkg/distributor/push_test.go:544]:
```suggestion
			err:                  httpgrpc.Error(http.StatusRequestEntityTooLarge, errMsg),
```

File: pkg/distributor/push_test.go [hunk 8]
```diff
@@ -590,29 +590,29 @@ func TestHandler_ErrorTranslation(t *testing.T) {
 		},
 		{
 			name:                 "a generic error during push gets a HTTP 500 without DoNotLogError header",
-			err:                  fmt.Errorf(errMsg),
+			err:                  fmt.Errorf("%s", errMsg),
```
Comment (aknuds1) [inline pkg/distributor/push_test.go:593]:
```suggestion
			err:                  errors.New(errMsg),
```

File: pkg/distributor/push_test.go [hunk 9]
```diff
@@ -590,29 +590,29 @@ func TestHandler_ErrorTranslation(t *testing.T) {
...
 			expectedHTTPStatus:   http.StatusInternalServerError,
 			expectedErrorMessage: errMsg,
 			expectedLogs:         []string{`level=error user=testuser msg="detected an error while ingesting Prometheus remote-write request (the request may have been partially ingested)" httpCode=500 err="this is an error"`},
 		},
 		{
 			name:                        "a DoNotLogError of a generic error during push gets a HTTP 500 with DoNotLogError header",
-			err:                         middleware.DoNotLogError{Err: fmt.Errorf(errMsg)},
+			err:                         middleware.DoNotLogError{Err: fmt.Errorf("%s", errMsg)},
```
Comment (aknuds1) [inline pkg/distributor/push_test.go:600]:
```suggestion
			err:                         middleware.DoNotLogError{Err: errors.New(errMsg)},
```

File: pkg/distributor/push_test.go [hunk 10]
```diff
@@ -590,29 +590,29 @@ func TestHandler_ErrorTranslation(t *testing.T) {
...
 			expectedHTTPStatus:          http.StatusInternalServerError,
 			expectedErrorMessage:        errMsg,
 			expectedDoNotLogErrorHeader: true,
 			expectedLogs:                []string{`level=error user=testuser msg="detected an error while ingesting Prometheus remote-write request (the request may have been partially ingested)" httpCode=500 err="this is an error"`},
 		},
 		{
 			name:                 "a gRPC error with a status during push gets translated into HTTP error without DoNotLogError header",
-			err:                  httpgrpc.Errorf(http.StatusRequestEntityTooLarge, errMsg),
+			err:                  httpgrpc.Errorf(http.StatusRequestEntityTooLarge, "%s", errMsg),
```
Comment (aknuds1) [inline pkg/distributor/push_test.go:608]:
```suggestion
			err:                  httpgrpc.Error(http.StatusRequestEntityTooLarge, errMsg),
```

File: pkg/distributor/push_test.go [hunk 11]
```diff
@@ -590,29 +590,29 @@ func TestHandler_ErrorTranslation(t *testing.T) {
...
 			expectedHTTPStatus:   http.StatusRequestEntityTooLarge,
 			expectedErrorMessage: errMsg,
 			expectedLogs:         []string{`level=error user=testuser msg="detected an error while ingesting Prometheus remote-write request (the request may have been partially ingested)" httpCode=413 err="rpc error: code = Code(413) desc = this is an error" insight=true`},
 		},
 		{
 			name:                        "a DoNotLogError of a gRPC error with a status during push gets translated into HTTP error without DoNotLogError header",
-			err:                         middleware.DoNotLogError{Err: httpgrpc.Errorf(http.StatusRequestEntityTooLarge, errMsg)},
+			err:                         middleware.DoNotLogError{Err: httpgrpc.Errorf(http.StatusRequestEntityTooLarge, "%s", errMsg)},
```
Comment (aknuds1) [inline pkg/distributor/push_test.go:615]:
```suggestion
			err:                         middleware.DoNotLogError{Err: httpgrpc.Error(http.StatusRequestEntityTooLarge, errMsg)},
```

File: pkg/scheduler/scheduler.go [hunk 12]
```diff
@@ -405,7 +405,7 @@ func (s *Scheduler) cancelRequestAndRemoveFromPending(key queue.RequestKey, reas
 
 	req := s.schedulerInflightRequests[key]
 	if req != nil {
-		req.CancelFunc(cancellation.NewErrorf(reason))
+		req.CancelFunc(cancellation.NewErrorf("%s", reason))
```
Comment (aknuds1) [inline pkg/scheduler/scheduler.go:408]:
```suggestion
		req.CancelFunc(cancellation.NewError(reason))
```

File: tools/tsdb-index-toc/main.go [hunk 13]
```diff
@@ -33,18 +33,18 @@ func main() {
 
 	finfo, err := os.Stat(filepath)
 	if err != nil {
-		log.Fatalf(err.Error())
+		log.Fatalf("%s", err.Error())
```
Comment (aknuds1) [inline tools/tsdb-index-toc/main.go:36]:
```suggestion
		log.Fatal(err.Error())
```

File: tools/tsdb-index-toc/main.go [hunk 14]
```diff
@@ -33,18 +33,18 @@ func main() {
...
 	}
 	indexSize := finfo.Size()
 
 	f, err := fileutil.OpenMmapFile(filepath)
 	if err != nil {
-		log.Fatalf(err.Error())
+		log.Fatalf("%s", err.Error())
```
Comment (aknuds1) [inline tools/tsdb-index-toc/main.go:42]:
```suggestion
		log.Fatal(err.Error())
```

File: tools/tsdb-index-toc/main.go [hunk 15]
```diff
@@ -33,18 +33,18 @@ func main() {
...
 	}
 
 	toc, err := index.NewTOCFromByteSlice(realByteSlice(f.Bytes()))
 	if err != nil {
-		log.Fatalf(err.Error())
+		log.Fatalf("%s", err.Error())
```
Comment (aknuds1) [inline tools/tsdb-index-toc/main.go:47]:
```suggestion
		log.Fatal(err.Error())
```

**Review by aknuds1 — CHANGES_REQUESTED** [review]
Please see suggestions for cleanup.

File: pkg/scheduler/scheduler.go (hunk 12 above)
Comment (alexweav) [reply pkg/scheduler/scheduler.go]: I had to reverse this one, as `reason` is a string, and `cancellation.NewError()` takes an inner error instead. It seems %s is still more appropriate here

Comment (aknuds1) [reply pkg/scheduler/scheduler.go:408]: I forgot it takes an error.

Comment (aknuds1) [reply pkg/scheduler/scheduler.go:408]: [Nit]
```suggestion
		req.CancelFunc(cancellation.NewError(errors.New(reason)))
```

**Review by aknuds1 — COMMENTED** [review]
Sorry, my suggestion got lost in the previous review.

**Comment by alexweav** [comment]
Indeed, I do wish the linter would report all the errors at once on the CI rather than reporting them in batches.

**Comment by aknuds1** [comment]
> Indeed, I do wish the linter would report all the errors at once on the CI rather than reporting them in batches.

It's the easiest to lint locally.

**Comment by alexweav** [comment]
Yeah, having to fall back to that it seems. I was misled by the last round (having only one error reported) into thinking I am done :see_no_evil:

**Review by aknuds1 — APPROVED** [review]
LGTM, thanks!

**Comment by alexweav** [comment]
Thanks for the review @aknuds1!

**Pull request merged**